### PR TITLE
Flux Monitor uses a single log subscription for all FluxAggregators

### PR DESCRIPTION
**Pull request by spooktheducks**

### CODE REVIEW

File: core/services/fluxmonitor/flux_monitor.go [hunk 1]
```diff
@@ -689,29 +592,29 @@ func (p *PollingDeviationChecker) createJobRun(nextPrice decimal.Decimal, nextRo
 var dec0 = decimal.NewFromInt(0)
 
 // OutsideDeviation checks whether the next price is outside the threshold.
-func OutsideDeviation(curPrice, nextPrice decimal.Decimal, threshold float64) bool {
-	if curPrice.Equal(dec0) {
-		logger.Infow("Current price is 0, deviation automatically met", "currentPrice", dec0)
+func OutsideDeviation(curAnswer, nextAnswer decimal.Decimal, threshold float64) bool {
+	if curAnswer.Equal(dec0) {
+		logger.Infow("Current price is 0, deviation automatically met", "answer", dec0)
```
Comment (RyanRHall) [inline core/services/fluxmonitor/flux_monitor.go]: should this maybe rethought and instead automatically poll only if the currentRound is 0? I think the purpose here is to automatically poll on the first round, when the current answer is 0... but in the future I imagine we may want to consider 0 values as potentially valid instead of trying to re-poll.

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 2]
```diff
@@ -674,6 +674,46 @@ contract FluxAggregator is AggregatorInterface, Owned {
     return currentRound.add(1);
   }
 
+  function roundState()
```
[the remark this answers is not in the document]
Comment (RyanRHall) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol]: Maybe check with @se3000 and/or @kaleofduty

File: core/services/fluxmonitor/flux_monitor.go [hunk 3]
```diff
@@ -187,76 +183,48 @@ func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
...
+	}
+	fm.removes <- *id
 }
 
 // DeviationCheckerFactory holds the New method needed to create a new instance
 // of a DeviationChecker.
 type DeviationCheckerFactory interface {
 	New(models.Initiator, RunManager, *orm.ORM) (DeviationChecker, error)
+	Start()
+	Stop()
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/fluxmonitor/flux_monitor.go]: Oops, management of the logBroadcaster should be in the parent, not the factory

File: core/services/fluxmonitor/flux_monitor.go (hunk 1 above)
Comment (spooktheducks) [reply core/services/fluxmonitor/flux_monitor.go]: Maybe so, but I left this logic identical to how I found it.  Maybe throw a story in tracker?

File: core/utils/utils.go [hunk 4]
```diff
@@ -515,3 +516,12 @@ func Uint256ToHex(n *big.Int) (string, error) {
 	}
 	return common.BigToHash(n).Hex(), nil
 }
+
+var dec10 = decimal.NewFromInt(10)
+
+func DecimalFromBigInt(i *big.Int, precision int32) decimal.Decimal {
+	return decimal.NewFromBigInt(i, -precision)
+	raw := decimal.NewFromBigInt(i, 0)
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/utils/utils.go]: Oops, meant to remove the stuff below the return

File: tools/ci/ethereum_test [hunk 5]
```diff
@@ -27,8 +27,6 @@ cd $SRCROOT/integration
 deploy_v05_contracts
 ./service_agreement_test
 
-deploy_v06_contracts
-./flux_monitor_test
```
Comment (RyanRHall) [inline tools/ci/ethereum_test]: If we remove this bash test we should probably make the `geth-postgres-ts` required for merging. I think @se3000 may have to do that? Also be aware that removing this affects both `geth-postgres` and `parity-postgres` and we currently don't have a `parity-postgres-ts` substitute.

Comment (spooktheducks) [reply tools/ci/ethereum_test]: Ryan is setting up the Parity environment in another PR.

File: core/services/fluxmonitor/flux_monitor.go (hunk 1 above)
Comment (RyanRHall) [reply core/services/fluxmonitor/flux_monitor.go]: Done.
https://www.pivotaltracker.com/story/show/171869391

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 6]
```diff
@@ -674,16 +674,61 @@ contract FluxAggregator is AggregatorInterface, Owned {
     return currentRound.add(1);
   }
 
+  function roundState(address _oracle)
+    external
+    view
+    returns (uint32 _reportableRoundId, bool _eligibleToSubmit, int256 _latestAnswer)
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol]: Unfortunately, we do -- it's the only way to get them to decode into Go struct fields using Geth's helpers.  Is it a problem that it shadows that function?

[the remark this answers is not in the document]
Comment (thodges-gh) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol]: Cool, yeah, that's the only reason, that it shadows a function. So we can just rename it.

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 7]
```diff
@@ -154,11 +154,11 @@ contract FluxAggregator is AggregatorInterface, Owned {
...
     onlyUnenabledAddress(_oracle)
   {
-    require(oracleCount() < 42, "cannot add more than 42 oracles");
-    require(_admin != address(0), "admin address must not be 0x0");
+    require(oracleCount() < 42);
+    require(_admin != address(0));
     require(
       oracles[_oracle].admin == address(0) || oracles[_oracle].admin == _admin,
-      "cannot modify previously-set admin address"
+      ""
```
Comment (RyanRHall) [inline evm-contracts/src/v0.6/dev/FluxAggregator.sol]: why did we remove the error messages here?

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 8]
```diff
@@ -360,15 +360,15 @@ contract FluxAggregator is AggregatorInterface, Owned {
...
-    return rounds[reportingRoundId].startedAt;
+  /**	
+   * @notice get the start time of the current reporting round	
+   */	
+  function reportingRoundStartedAt()	
+    external	
+    view	
+    returns (uint256)	
+  {	
+    return rounds[reportingRoundId].startedAt;
```
Comment (RyanRHall) [inline evm-contracts/src/v0.6/dev/FluxAggregator.sol]: :scissors: added whitespace

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol (hunk 7 above)
Comment (RyanRHall) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol]: And if they're going to stay removed, we should be able to omit the empty string.

File: tools/ci-ts/tests/flux-monitor.test.ts [hunk 9]
```diff
@@ -9,7 +9,7 @@ import {
 } from '../test-helpers/common'
 import * as clClient from '../test-helpers/chainlink-cli'
 import { contract, helpers as h, matchers } from '@chainlink/test-helpers'
-import { FluxAggregatorFactory } from '@chainlink/contracts/ethers/v0.6/FluxAggregatorFactory'
+import { FluxAggregatorFactory } from '../../../evm-contracts/ethers/v0.6/FluxAggregatorFactory'
```
Comment (RyanRHall) [inline tools/ci-ts/tests/flux-monitor.test.ts]: why was this changed? I think it's cleaner to use the workspace names in imports.

File: tools/ci-ts/tests/flux-monitor.test.ts [hunk 10]
```diff
@@ -105,6 +105,7 @@ describe('flux monitor eth client integration', () => {
     fluxMonitorJob.initiators[0].params.address = fluxAggregator.address
     job = clClient.createJob(JSON.stringify(fluxMonitorJob))
     assert.equal(clClient.getJobs().length, initialJobCount + 1)
+    await wait(10000)
```
Comment (RyanRHall) [inline tools/ci-ts/tests/flux-monitor.test.ts]: we shouldn't have to artificially wait here because the `assertJobRun` func already asserts over a period of 20 seconds

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol (hunk 7 above)
Comment (spooktheducks) [reply evm-contracts/src/v0.6/dev/FluxAggregator.sol]: Error messages put us over the contract size limit.  We're looking into better fixes in another PR, so this is hopefully temporary.  I'll remove this empty string, just an oversight.

File: tools/ci-ts/tests/flux-monitor.test.ts (hunk 9 above)
Comment (spooktheducks) [reply tools/ci-ts/tests/flux-monitor.test.ts]: As far as I could tell, it wasn't pulling changes I was making to that package because I hadn't published those to `npm`.  This way, the changes can all be part of a single PR, and we can publish as soon as it's accepted.

File: core/eth/client.go [hunk 11]
```diff
@@ -29,6 +23,7 @@ type Client interface {
 	GetERC20Balance(address common.Address, contractAddress common.Address) (*big.Int, error)
 	SendRawTx(hex string) (common.Hash, error)
 	GetTxReceipt(hash common.Hash) (*TxReceipt, error)
+	GetBlockHeight() (uint64, error)
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/eth/client.go]: It's called when the PollingDeviationChecker spins up

File: core/eth/types.go [hunk 12]
```diff
@@ -58,6 +58,26 @@ func (log Log) GetTopic(idx uint) (common.Hash, error) {
 	return log.Topics[idx], nil
 }
 
+func (log Log) Copy() Log {
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/eth/types.go]: Sure -- there's one in log_broadcaster.go where this method is called, but I can add one over here as well.

File: core/eth/client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (se3000) [reply core/eth/client.go]: Would it make sense to read that from the db instead? That's the state the rest of the app is at.

[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/eth/client.go]: Actually, it looks like that change got swallowed somewhere... good catch.

File: tools/ci-ts/tests/flux-monitor.test.ts (hunk 9 above)
Comment (RyanRHall) [reply tools/ci-ts/tests/flux-monitor.test.ts]: It should incorporate your changes without having to publish anything. The namespaces are automatically resolved. Maybe you were missing a build step? `yarn setup:contracts`

File: integration-scripts/src/deployV0.6Contracts.ts [hunk 13]
```diff
@@ -25,7 +25,7 @@ export async function deployContracts() {
     { Factory: FluxAggregatorFactory, signer, name: 'FluxAggregator' },
     linkToken.address,
     1,
-    3,
+    60,
```
Comment (se3000) [inline integration-scripts/src/deployV0.6Contracts.ts]: What is this change about?

File: integration/flux_monitor_test [hunk 14]
```diff
@@ -8,7 +8,7 @@ flux_monitor_test() {
   local log=$LOG_PATH/flux_monitor.log
   local expected_job_count=$(expr $(chainlink -j jobs list | jq length) + 1)
 
-  local pollingInterval="5s"
+  local pollingInterval="8s"
```
Comment (se3000) [inline integration/flux_monitor_test]: Why bump this?

File: tools/ci/ethereum_test [hunk 15]
```diff
@@ -27,8 +27,6 @@ cd $SRCROOT/integration
...
```
Comment (se3000) [inline tools/ci/ethereum_test]: Can we remove this code entirely now?

Comment (RyanRHall) [reply tools/ci/ethereum_test]: It's removed in https://github.com/smartcontractkit/chainlink/pull/2524. Can you rebase?

File: integration-scripts/src/deployV0.6Contracts.ts (hunk 13 above)
Comment (spooktheducks) [reply integration-scripts/src/deployV0.6Contracts.ts]: This raises the round timeout to 60 seconds so that we get more deterministic test behavior, particularly if the CI node happens to be under load.

File: tools/ci/ethereum_test (hunk 15 above)
Comment (spooktheducks) [reply tools/ci/ethereum_test]: Definitely.  Ryan's TS integration test covers the old Bash stuff 100%.

File: integration/flux_monitor_test (hunk 14 above)
Comment (spooktheducks) [reply integration/flux_monitor_test]: (removing this test entirely)
